### PR TITLE
feat(config): Add hooks

### LINKED ISSUE #902: Per source hooks in the config file
Thanks a lot for creating wonderful alternative to the amazing `restic` program. I have been watching the progress of your project for a while and waiting for the right moment to jump on. We a few things have happened that cause me to seriously take a look at this project. I have been looking to restore a very old snapshot in restic and found the lack of any description very challenging in figuring out the right snapshot to pick. So the fact that this one has that ability is great incentive going forward. However I am using a wrapper for restic called autorestic which has the ability to call scripts/programs per source/location as hooks. I use this ability extensively to gracefully bring down containers before backup and bring them back up after the backup is complete. There is a documentation available for this feature [here](https://autorestic.vercel.app/location/hooks). There are 4 hooks supported:
- before
- after
- failure
- success

These hooks are per source. But I think there might be some value in having hooks that are called per repo as well.

### PR DESCRIPTION
**Pull request by aawsome**
This PR adds `run-before`, `run-after`, `run-failure` and `run-finally` hooks for:
- all commands in the `[global.hooks]` config profile section
- commands accessing the repository in the `[repository.hooks]` config profile section
- the `backup` command specifically in the `[backup.hooks]` config profile section
- specific backup sources in the `[backup.snapshots.hooks]` section

Note: This PR includes only calling the given commands. If there is the wish for supplying information to the commands (env variables or parameter substitution), this should be covered by a separate feature request/PR.

closes #902

### CODE REVIEW

File: src/commands/backup.rs [hunk 1]
```diff
@@ -264,7 +280,9 @@ impl BackupCmd {
             }
 
             info!("backup of {source} done.");
+            run_after.run(&format!("run-after backup {source}"))?;
         }
+        config.backup.run_after.run("run-after backup")?;
```
Comment (nardoor) [inline src/commands/backup.rs:285]: do we always want to run `run-after` even if somewhere fails after a `run-before`? 
in an analog way of `open` / `close` file handles?

File: src/commands/backup.rs [hunk 2]
```diff
@@ -221,9 +232,14 @@ impl BackupCmd {
...
             }
 
+            // use the correct source-specific run-before and run-after
+            let run_before = opts.run_before.clone();
+            let run_after = opts.run_after.clone();
+
             // merge "backup" section from config file, if given
             opts.merge(config.backup.clone());
 
+            run_before.run(&format!("run-before backup {source}"))?;
```
Comment (nardoor) [inline src/commands/backup.rs:242]: I feel like `rustic` executing commands could almost always do an `info` log.
To me, because the commands executed depend on both the user and `rustic`, the user should be informed of what rustic does with the provided commands.
In the `CommandInput` type for instance, or before calling the command.

File: src/config.rs [hunk 3]
```diff
@@ -242,3 +251,58 @@ fn get_global_config_path() -> Option<PathBuf> {
...
     Some(PathBuf::from("/etc/rustic"))
 }
+
+/// A command to be called which can be given as CLI option as well as in config files
+/// `CommandInput` implements Serialize/Deserialize as well as FromStr.
+
+// Note: we use CommandInputInternal here which itself impls FromStr in order to use serde_as PickFirst for CommandInput.
+#[serde_as]
+#[derive(Debug, Default, Clone, Serialize, Deserialize, Merge)]
+pub struct CommandInput(#[serde_as(as = "PickFirst<(_,DisplayFromStr)>")] CommandInputInternal);
```
Comment (nardoor) [inline src/config.rs:261]: If we want to backport this dual compatibility (`String` and `Vec<String>`) for commands in `rustic_core` crate (regarding the changes we made a few weeks ago concerning `password` commands, I assume this `CommandInput` code might be moved to `rustic_core` crate is that right ?

**Review by nardoor — COMMENTED** [review]
I feel like this feature deserves: 
- an integration test in some way
- a page in the cargo book docs

File: src/commands/backup.rs [hunk 4]
```diff
@@ -151,6 +160,8 @@ impl BackupCmd {
         }
         .to_indexed_ids()?;
 
+        config.backup.run_before.run("run-before backup")?;
+
         // manually check for a "source" field, check is not done by serde, see above.
         if !config.backup.source.is_empty() {
             bail!("key \"source\" is not valid in the [backup] section!");
```
[the remark this answers is not in the document]
Comment (aawsome) [reply src/commands/backup.rs:167]: Good catch! I's change this.

File: src/commands/backup.rs (hunk 1 above)
Comment (aawsome) [reply src/commands/backup.rs:285]: I think there are cases. For instance, if users want to mount a filesystem before backing it up - then they probably want to unmount mo-matter-what.
But actually this is also a weakness - if we catch an error in between, the `run-after` command won't even be executed...

I think adding additional hooky like `run-after-failed` or `run-after-successful` should be the solution here. I'll see if we can add them here..

File: src/commands/backup.rs (hunk 2 above)
Comment (aawsome) [reply src/commands/backup.rs:242]: Oh - this is always a discussion between people wanting more and people wanting less verbosity... 
I wouldn't log in `info` in the `CommandInput` as in `rustic_core` there might be situations where such a command is called quite often...
Currently this is a `trace` - we could discuss if `debug` is better suited..

File: src/config.rs (hunk 3 above)
Comment (aawsome) [reply src/config.rs:261]: Exactly.

File: src/commands/backup.rs (hunk 2 above)
Comment (nardoor) [reply src/commands/backup.rs:242]: `debug` could be a good compromise indeed

File: src/commands/backup.rs (hunk 1 above)
Comment (nardoor) [reply src/commands/backup.rs:285]: Should `run-after`, `run-after-success`, `run-after-failed` exist at the same time?
Or do you think `run-after-success`, `run-after-failed` can replace and exist without ~run-after~?

Comment (aawsome) [reply src/commands/backup.rs]: I actually called them `run-failed` and `run-finally`.
- `run-after` is only called if everything was ok
- `run-failed` is only called if there was a failure
- `run-finally` is always called

File: src/commands/backup.rs (hunk 2 above)
Comment (aawsome) [reply src/commands/backup.rs]: It is now `debug`

File: src/config.rs [hunk 5]
```diff
@@ -252,13 +250,75 @@ fn get_global_config_path() -> Option<PathBuf> {
...
+#[serde(default, rename_all = "kebab-case", deny_unknown_fields)]
+pub struct Hooks {
+    /// Call this command before every rustic operation
+    pub run_before: CommandInput,
+
+    /// Call this command after every successful rustic operation
+    pub run_after: CommandInput,
+
+    /// Call this command after every rustic operation
+    pub run_failed: CommandInput,
```
Comment (zxcat) [inline src/config.rs:263]: typo in comment: "every" → "every failed"

File: src/config.rs (hunk 3 above)
Comment (aawsome) [reply src/config.rs]: This PR is now based on https://github.com/rustic-rs/rustic_core/pull/252

File: src/config.rs (hunk 5 above)
Comment (aawsome) [reply src/config.rs:261]: Good catch, thanks! Is fixed now.

File: Cargo.toml [hunk 6]
```diff
@@ -41,8 +41,8 @@ rustdoc-args = ["--document-private-items", "--generate-link-to-definition"]
 
 [dependencies]
 abscissa_core = { version = "0.7.0", default-features = false, features = ["application"] }
-rustic_backend = { version = "0.2.1", features = ["cli"] }
-rustic_core = { version = "0.3.1", features = ["cli"] }
+rustic_backend = { git = "https://github.com/rustic-rs/rustic_core.git", features = ["cli"] }
+rustic_core = { git = "https://github.com/rustic-rs/rustic_core.git", features = ["cli"] }
```
Comment (simonsan) [inline Cargo.toml:45]: I think we may want to release the changes first without re-introducing git dependencies again, changing them back later. Now that releasing should be just merging the release pr.

Comment (aawsome) [reply Cargo.toml:45]: If we wait for a release of rustic_core, we'll pile up PRs in rustic which depend on these things but cannot be merged. We won't have the features in the nightly builds and will not get any feedback until we release a rustic release.

As many changes in rustic are changes in rustic_core, I see no alternative than using a git dependency here.

Comment (simonsan) [reply Cargo.toml:45]: Ok, I understand. Do you think it would be viable to up the release cadence for `rustic_core` and say we do at least one release in a week with things that have been merged, even if small one's? It would be just a click away, we could release another `rustic_core` version today, for example.

File: src/commands/backup.rs [hunk 7]
```diff
@@ -206,82 +211,134 @@ impl BackupCmd {
...
-            info!("backup of {source} done.");
+        if is_err {
+            bail!("Not all sources were successfully backuped!")
         }
 
         Ok(())
     }
 }
+
+fn backup_source<P: ProgressBars, S: IndexedIds>(
```
Comment (nardoor) [inline src/commands/backup.rs:253]: couldn't/shouldn't this be part of an `impl BackupCmd` ?

something ~close to that?
```rust
impl BackupCmd {
   fn backup_source<P: Pr.., S: Inde..(&mut self, source: PathList, repo: &Repository<P, S>) 
}
```

**Comment by nardoor** [comment]
Just wondering:
I get that `hooks` can be **global**.
I see how `hooks` in `BackupCmd` can allow per-source `hooks`.

But I fail to think of a use case for `hooks` in `CopyCmd`. Could you elaborate about that for me about that @aawsome pls?

**Comment by aawsome** [comment]
@nardoor A use case could be:
```
[repository]
repository = "/mnt/repo/

[global.hooks]
run-before = ["mount /mnt"]
run-finally = ["umount /mnt"]
```

which would simply mount an umount a (maybe external) drive before and after each rustic operation. Now, when copying to such a repository, we would like to run the hooks too.

File: src/commands/copy.rs [hunk 8]
```diff
@@ -89,69 +91,88 @@ impl CopyCmd {
...
-            };
-
-            if poly != repo_dest.config().poly()? {
-                bail!("cannot copy to repository with different chunker parameter (re-chunking not implemented)!");
+            let hooks = target_config.global.hooks.with_context("target");
+            if let Err(err) = self.copy(&repo, target_opt, &snapshots, &hooks) {
+                error!("error copying to target: {err}");
+                hooks.run_failed()?;
             }
+            hooks.run_finally()?;
```
Comment (nardoor) [inline src/commands/copy.rs:99]: My understanding: we apply "global" hooks for a 2nd (target) repository.
But we do it from the command implementation. I am thinking this a bit tricky, as the link between "opening a 2nd repo with some `target` config" and "applying the global hooks from the `target` config" exist in the `CopyCmd` implementation.

In an ideal world, this link would rather be defined in some `type` or factorized `function` that handles anything related with an extra `target` repository.

My fear: tomorrow someone implements some other command that interacts with a 2nd `target` repo and forgets to use the `global.hooks` which could lead to a bug.

I might be over-anticipating, let me know what you about your opinion on that!

File: src/commands/backup.rs (hunk 7 above)
Comment (aawsome) [reply src/commands/backup.rs]: is changed.

File: src/commands/copy.rs (hunk 8 above)
Comment (aawsome) [reply src/commands/copy.rs:99]: Yes, you are perfectly right!
Actually the `copy` command so far is the only use case for multiple repositories, but there could be more.

So far the global hooks are not implemented when the repository is opened or closed, which should be the right place...

What do you think about adding even one more hooks section: `[repository.hooks]`. These would be hooks which are run when accessing a repository. In contrast, the `[global.hooks]` would be run for every command (even `self-update` or `show-config`). 

In this setting `[global.hooks]` would be mainly used to start log-like services, e.g. sending an email for failed runs, whereas `[repository.hooks]` would be used to define pre- and post-actions for repository access.

Then, the `copy` command would only call the `[repository.hooks]`.

**Comment by aawsome** [comment]
I rebased this to main (merging main into it was getting too complicated)..

I'd like to add `[repository.hooks]` as suggested - I think I can do it this evening. Then this PR will  be ready, from my side.

File: config/README.md [hunk 9]
```diff
@@ -65,6 +65,17 @@ If you want to contribute your own configuration, please
 | progress-interval | The interval at which progress indicators are shown.                              | "100ms"       | "1m"                     | RUSTIC_PROGRESS_INTERVAL |
 | use-profile       | Profile or array of profiles to use. Allows to recursely use other profiles.      | Empty array   | "other" , ["2nd", "3rd"] | RUSTIC_USE_PROFILE       |
 
+### Global Hooks `[global.hooks]`
+
+These external commands are before and after each commands, respectively.
```
Comment (simonsan) [inline config/README.md:70]:
```suggestion
These external commands are run before and after each commands, respectively.
```

File: config/README.md [hunk 10]
```diff
@@ -65,6 +65,17 @@ If you want to contribute your own configuration, please
...
+
+| Attribute   | Description                                      | Default Value | Example Value                 | Environment Variable |
+| ----------- | ------------------------------------------------ | ------------- | ----------------------------- | -------------------- |
+| run-before  | Run the given command before execution           | not set       | "echo test", ["echo", "test"] |                      |
+| run-after   | Run the given command after successful execution | not set       | "echo test", ["echo", "test"] |                      |
+| run-failed  | Run the given command after failed execution     | not set       | "echo test", ["echo", "test"] |                      |
+| run-finally | Run the given command after every execution      | not set       | "echo test", ["echo", "test"] |                      |
+
```
Comment (simonsan) [inline config/README.md:78]:
```suggestion
| run-before  | Run the given command before execution           | not set       | ["echo test"] |                      |
| run-after   | Run the given command after successful execution | not set       | ["echo test"] |                      |
| run-failed  | Run the given command after failed execution     | not set       | ["echo test"] |                      |
| run-finally | Run the given command after every execution      | not set       | ["echo test"] |                      |

```

File: config/README.md [hunk 11]
```diff
@@ -166,11 +177,26 @@ can be overwritten in the source-specifc configuration, see below.
 | time                  | Set the time saved in the snapshot.                                                     | Not set               |               |
 | with-atime            | If true, includes file access time (atime) in the backup.                               | false                 |               |
 
+### Backup Hooks `[backup.hooks]`
+
+These external commands are before and after each backup, respectively. Note
+that global hooks are run additionaly.
```
Comment (simonsan) [inline config/README.md:183]:
```suggestion
These external commands are run before and after each backup, respectively.

**Note**: Global hooks are run additionally.
```

File: config/README.md [hunk 12]
```diff
@@ -166,11 +177,26 @@ can be overwritten in the source-specifc configuration, see below.
...
+
+| Attribute   | Description                                      | Default Value | Example Value                 | Environment Variable |
+| ----------- | ------------------------------------------------ | ------------- | ----------------------------- | -------------------- |
+| run-before  | Run the given command before execution           | not set       | "echo test", ["echo", "test"] |                      |
+| run-after   | Run the given command after successful execution | not set       | "echo test", ["echo", "test"] |                      |
+| run-failed  | Run the given command after failed execution     | not set       | "echo test", ["echo", "test"] |                      |
+| run-finally | Run the given command after every execution      | not set       | "echo test", ["echo", "test"] |                      |
```
Comment (simonsan) [inline config/README.md:190]:
```suggestion
| run-before  | Run the given command before execution           | not set       | ["echo test"] |                      |
| run-after   | Run the given command after successful execution | not set       | ["echo test"] |                      |
| run-failed  | Run the given command after failed execution     | not set       | ["echo test"] |                      |
| run-finally | Run the given command after every execution      | not set       | ["echo test"] |                      |
```

File: config/README.md [hunk 13]
```diff
@@ -166,11 +177,26 @@ can be overwritten in the source-specifc configuration, see below.
...
+
 ### Backup Sources `[[backup.sources]]`
 
 **Note**: All of the backup options mentioned before can also be used as
 source-specific option and then only apply to this source.
 
+Source-specific hooks are called when backup up the defined source, additionally
+to global and backup hooks. only apply
```
Comment (simonsan) [inline config/README.md:198]:
```suggestion
Source-specific hooks are called additionally to global and backup hooks
when backing up the defined source.

```

File: src/commands/backup.rs [hunk 14]
```diff
@@ -209,86 +214,138 @@ impl BackupCmd {
...
-                );
-                debug!("Data Blobs:  {} new", summary.data_blobs);
-                debug!("Tree Blobs:  {} new", summary.tree_blobs);
-                println!(
-                    "Added to the repo: {} (raw: {})",
-                    bytes_size_to_string(summary.data_added_packed),
-                    bytes_size_to_string(summary.data_added)
-                );
+        if is_err {
+            bail!("Not all sources were successfully backuped!")
```
Comment (simonsan) [inline src/commands/backup.rs:249]:
```suggestion
            bail!("Not all sources were successfully backed up!")
```

**Review by simonsan — COMMENTED** [review]
First iteration for text grammar in docs. Will review again, when the other changes have been pushed.

File: src/commands/copy.rs (hunk 8 above)
Comment (aawsome) [reply src/commands/copy.rs]: [repository.hooky] is now added.

File: config/README.md [hunk 15]
```diff
@@ -166,11 +187,22 @@ can be overwritten in the source-specifc configuration, see below.
...
+
+See Global Hooks.
+
 ### Backup Sources `[[backup.sources]]`
 
 **Note**: All of the backup options mentioned before can also be used as
 source-specific option and then only apply to this source.
 
+Source-specific hooks are called additionally to global, repository and backup
+hooks when backup up the defined source.
```
Comment (simonsan) [inline config/README.md:204]:
```suggestion
hooks when backing up the defined source.
```

File: src/commands/backup.rs [hunk 16]
```diff
@@ -208,87 +217,116 @@ impl BackupCmd {
...
+        mut self,
+        source: PathList,
+        repo: &Repository<P, S>,
+    ) -> Result<()> {
+        let config = RUSTIC_APP.config();
+        let snapshot_opts = &config.backup.snapshots;
+        if let Some(path) = &self.as_path {
+            // as_path only works in combination with a single target
+            if source.len() > 1 {
+                bail!("as-path only works with a single target!");
```
Comment (simonsan) [inline src/commands/backup.rs:258]:
```suggestion
                bail!("as-path only works with a single target profile!");
```

File: config/README.md [hunk 17]
```diff
@@ -65,6 +65,20 @@ If you want to contribute your own configuration, please
 | progress-interval | The interval at which progress indicators are shown.                              | "100ms"       | "1m"                     | RUSTIC_PROGRESS_INTERVAL |
 | use-profile       | Profile or array of profiles to use. Allows to recursely use other profiles.      | Empty array   | "other" , ["2nd", "3rd"] | RUSTIC_USE_PROFILE       |
 
+### Global Hooks `[global.hooks]`
+
+These external commands are before and after each commands, respectively.
+
+** Note **: There are also repository hooks, which should be used for commands
```
Comment (simonsan) [inline config/README.md:72]:
```suggestion
**Note**: There are also repository hooks, which should be used for commands
```

**Review by simonsan — COMMENTED** [review]
Small nits regarding docs

File: src/commands/backup.rs [hunk 18]
```diff
@@ -208,87 +217,116 @@ impl BackupCmd {
...
-                        info!("merging source=\"{path}\" section from config file");
-                        opts.merge(snapshot_opts[idx].clone());
-                    }
+                if let Err(err) = opts.backup_snapshot(sources.clone(), &repo) {
+                    error!("error backing up {sources}: {err}");
+                    is_err = true;
                 }
             }
+            if is_err {
+                Err(anyhow!("Not all snapshots were successfully generated!"))
```
Comment (simonsan) [inline src/commands/backup.rs:241]:
```suggestion
                Err(anyhow!("Not all snapshots were generated successfully!"))
```

File: src/commands/backup.rs (hunk 16 above)
Comment (simonsan) [reply src/commands/backup.rs:258]: My thinking is, that it is a user-facing error and it might make sense to stay consistent to talk about `target profiles` rather than `targets` (which is probably more understandable for us devs).

Comment (aawsome) [reply src/commands/backup.rs:258]: Actually, this is not the target as used in `copy` here. But the target path set by `--as-path` for a source path. But the message is wrong anyway: It should be: `as-path only works with a single source`! I'll change it.

**Review by simonsan — DISMISSED** [review]

**Comment by simonsan** [comment]
Needs conflicts to be resolved, then we can merge!

File: src/application.rs [hunk 19]
```diff
@@ -113,6 +118,12 @@ impl Application for RusticApp {
 impl RusticApp {
     /// Shut down this application gracefully, exiting with given exit code.
     fn shutdown_with_exitcode(&self, shutdown: Shutdown, exit_code: i32) -> ! {
```
Comment (simonsan) [inline src/application.rs:120]: We're not using that throughout the commands, right? So these hooks will not be run, I assume?

Comment (simonsan) [reply src/application.rs:120]: Ah no, we always call it, didn't see the changes in `Application::shutdown`

**Comment by simonsan** [comment]
## Possible testing strategy generated by ChatGPT (inspirational)

### 1. **Unit Testing Strategy**

Use mocks to simulate the commands and hook executions without relying on actual system changes.

#### a. **Global Hooks**

- **Test Global Hooks on All Commands**  
      - Verify that the `global.hooks` are executed for all commands, including ones that don’t interact with repositories (e.g., `docs`, `self-update`, etc.).  
      - **Scenarios to test**:  
        - Hooks in `run-before`, `run-after`, `run-failed`, and `run-finally` are executed at the appropriate times.  
        - Test the behavior when hooks fail (ensure `run-failed` and `run-finally` hooks are called correctly).  
      - **Commands to test**:  
        - Non-repository related commands: `docs`, `self-update`.  
      - **Test cases**:  
        1. **run-before**: Ensure the command doesn't execute if the hook fails.  
        2. **run-after**: Ensure the command executes successfully, and `run-after` is called.  
        3. **run-failed**: Simulate a failure during the command and verify the `run-failed` hook is executed.  
        4. **run-finally**: Ensure the `run-finally` hook is always executed, regardless of success or failure.

#### b. **Repository Hooks**

- **Test Repository Hooks on Repository Commands**  
      - Verify that the `repository.hooks` are executed for commands that interact with repositories (e.g., `backup`, `repoinfo`, etc.).  
      - Ensure these hooks **do not run** for commands that do not involve a repository (e.g., `doc`, `self-update`).  
      - **Commands to test**:  
        - Repository related commands: `backup`, `repoinfo`.  
        - Non-repository related commands: `docs`, `self-update`.  
      - **Test cases**:  
        1. Ensure `repository.hooks` are skipped for `docs` and `self-update`.  
        2. Ensure `repository.hooks` are executed before and after `backup` and `repoinfo` commands.  
        3. Test with failure scenarios and ensure `run-failed` is triggered.

#### c. **Backup-Specific Hooks**

- **Test Backup-Specific Hooks**  
      - Ensure that `backup.hooks` are only executed for the `backup` command.  
      - **Test cases**:  
        1. Verify `backup.hooks.run-before` is executed before starting a backup process.  
        2. Ensure `backup.hooks.run-after` is executed once the backup finishes successfully.  
        3. Simulate a backup failure and check that `backup.hooks.run-failed` is triggered.  
        4. Ensure `backup.hooks.run-finally` is executed in both success and failure cases.

#### d. **Invalid Hook Configurations**

- Test scenarios where invalid configurations are provided for hooks (e.g., missing commands, wrong paths, or malformed entries).  
      - Verify that proper errors are raised and the system behaves gracefully.

### 2. **Integration Testing Strategy**

#### a. **Test Command Flow with Hooks**

- Execute actual commands (e.g., `backup`, `repoinfo`, `docs`) and ensure the hooks are triggered in the right sequence:  
      - Global hooks should run on every command.  
      - Repository hooks should only run for repository-related commands.  
      - Backup hooks should only run for the backup command.

#### b. **Check Hook Output and Side Effects**

- Verify the side effects of each hook (e.g., logs, file creation, external commands executed).  
      - You can mock or spy on external commands invoked by hooks and check their execution.  
      - Verify that hooks write logs or produce expected output when executed.

#### c. **Test Failures and Recovery Mechanisms**

- Simulate scenarios where hooks or commands fail to ensure the system recovers properly:  
      - **Command fails after run-before**: Ensure `run-failed` and `run-finally` are executed.  
      - **Command succeeds but run-after fails**: Ensure `run-finally` is still executed.  
      - Ensure that the error handling doesn’t break the overall system (e.g., it doesn’t leave processes running, logs are written, etc.).

### 3. **Edge Cases and Negative Testing**

- **Empty Hook Arrays**: Test that no hooks are executed when arrays are empty (i.e., `[global.hooks.run-before]` has no effect if empty).
- **Multiple Hooks**: Test behavior with multiple hooks defined for the same phase (e.g., multiple `run-before` hooks). Ensure all hooks are executed in the right order.
- **Hooks That Fail**: Test cases where hooks themselves fail (e.g., invalid commands in the hook, failure in execution). Ensure proper logging and fallback mechanisms (e.g., `run-failed` and `run-finally` should still execute).

### 4. **Stress and Performance Testing**

- If hooks are complex and involve heavy tasks (e.g., interacting with external services, copying files), test the performance impact.  
      - Ensure hooks don’t cause significant performance degradation in the overall backup tool.  
      - Test with a large number of hooks to ensure they are executed efficiently.

### 5. **Documentation and User Feedback**

- Test that appropriate feedback is given to users (e.g., error messages if hooks fail).
- Ensure logs are written when hooks are executed, so users can verify the hook execution steps.

**Review by simonsan — APPROVED** [review]
LGTM!
